sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md

```diff
@@ -0,0 +1,776 @@
+  template <typename EltTy> auto format();
+  template <typename EltTy> auto format() &&;
+  template <typename EltTy, int Height, int Width> auto format() &;
+  template <typename EltTy, int Height, int Width> auto format() &&;
+
+  // Merge operations.
+  void merge(const value_type &Val, const mask_type_t<length> &Mask);
+  void merge(const value_type &Val1, value_type Val2,
+             const mask_type_t<length> &Mask);
+} // namespace gpu
+} // namespace intel
+} // namespace sycl
+```
+
+```sycl::intel::gpu::simd_view``` class supports all the element-wise operations and
+other utility functions defined for ```sycl::intel::gpu::simd``` class. It also
+provides region accessors and more generic operations tailored for 2D regions,
+such as row/column operators and 2D select/replicate/format/merge operations.
+
+```cpp
+  simd<float, 32> v1;
+  auto m1 = v1.format<float, 4, 8>();
+  simd<float, 4> v2;
+  auto m2 = v2.format<float, 2, 2>();
+
+  // ...
+  m2 = m1.select<2, 2, 2, 4>(1, 2);  // v_size = 2, v_stride = 2,
+                                     // h_size = 2, h_stride = 4,
+                                     // v_offset = 1, h_offset = 2.
+```
+<p align="center">
+<img src="images/Matrix_2_2_2_4__1_2.svg" title="2D select example" width="700" height="200"/>
+</p>
+
+```cpp
+  m1.select<4, 1, 4, 2>(0, 0) = 0.0f; // selected elements of m1
+                                      // are replaced with 0.0f
+```
+
+<p align="center">
+<img src="images/Matrix_4_1_4_2__0_0.svg" title="2D select example" width="700" height="200"/>
+</p>
+
+### Reduction functions
+
+Explicit SIMD provides the following reduction functions for simd objects.
+Compiler will produce optimal code sequence on the target device to apply the
+specified operation to all scalar elements in the input simd vector. Note that
+the order of element-wise operations is not guaranteed and the correctness of
+result should not depend on a particular computation order.
+- ```reduce(simd<T1, SZ> v, std::plus<>())```: returns a scalar value of type
+**T0** that's equal to the sum of all data elements of type **T1** in simd vector
+**v** of length **SZ**.
+- ```reduce(simd<T1, SZ> v, std::multiplies<>())```: returns a scalar value of
+type **T0** that's equal to the product of all data elements of type **T1** in simd
+vector **v** of length **SZ**.
+- ```hmax(simd<T1, SZ> v)```: returns a scalar value of type **T0**
+that's equal to the maximum value of all data elements of type **T1** in simd
+vector **v** of length **SZ**.
+- ```hmin(simd<T1, SZ> v)```: returns a scalar value of type **T0**
+that's equal to the minimum value of all data elements of type **T1** in simd
+vector **v** of length **SZ**.
+
+```cpp
+template <typename T0, typename T1, int SZ, typename BinaryOperation>
+T0 reduce(simd<T1, SZ> v, BinaryOperation op);
+
+template <typename T0, typename T1, int SZ> T0 hmax(simd<T1, SZ> v);
+template <typename T0, typename T1, int SZ> T0 hmin(simd<T1, SZ> v);
```

Comment:
What would be `vmax` then? Does it change if we turn the processor? :-)
Would `inner_max` or something like that would make sense?
I do not know whether @mattkretz looked at these kind of operations and studied the naming.

Reply:
The names `hmax` and `hmin` were discussed at some point in the Parallelism TS 2 process. I don't remember exactly, but I'm pretty sure there wasn't much concern. I agree the names are not perfect. (sounds like useful [TS feedback](https://github.com/mattkretz/std-simd-feedback/issues/16) :wink: ) `inner_min/max` sounds too mathy IMHO but still a bit more obvious. `reduce_max` or `max_reduce` are also not obviously great. (`valarray` has member functions `min()` and `max()`. But member functions inhibit writing generic code that also works with arithmetic types.) `min/max_element(simd)` maybe? In any case, would be great to align with what (hopefully) goes into C++23.